src/Microsoft.AspNet.SignalR.ServiceBus3/ServiceBusConnection.cs

```diff
@@ -0,0 +1,326 @@
+
+        [SuppressMessage("Microsoft.Design", "CA1031:DoNotCatchGeneralExceptionTypes", Justification = "We retry to create the topics on exceptions")]
+        private void Retry(Action action)
+        {
+            string errorMessage = "Failed to create service bus subscription or topic : {0}";
+            while (true)
+            {
+                try
+                {
+                    action();
+                    break;
+                }
+                catch (UnauthorizedAccessException ex)
+                {
+                    _trace.TraceError(errorMessage, ex.Message);
+                    break;
+                }
+                catch (MessagingException ex)
+                {
+                    _trace.TraceError(errorMessage, ex.Message);
+                    if (ex.IsTransient)
+                    {
+                        Thread.Sleep(RetryDelay);
+                    }
+                    else
+                    {
+                        break;
+                    }
+                }
+                catch (Exception ex)
+                {
+                    _trace.TraceError(errorMessage, ex.Message);
+                    Thread.Sleep(RetryDelay);
+                }
+            }
+        }
+
+        protected virtual void Dispose(bool disposing)
+        {
+            if (disposing)
+            {
+                // Close the factory
+                if (_factory != null)
+                {
+                    _factory.Close();
+                }
+            }
+        }
+
+        public void Dispose()
+        {
+            Dispose(true);
+        }
+
+        private async Task ProcessMessages(ReceiverContext receiverContext)
+        {
+            TimeSpan backoffAmount = _backoffTime;
+
+            while (true)
+            {
+                try
+                {
+                    var messages = await receiverContext.Receiver.ReceiveBatchAsync(
+                        ReceiverContext.ReceiveBatchSize, receiverContext.ReceiveTimeout);
+
+                    receiverContext.OnMessage(messages);
+
+                    // Reset the receive timeout if it changed
+                    receiverContext.ReceiveTimeout = DefaultReadTimeout;
+
+                    continue;
+                }
+                catch (OperationCanceledException)
+                {
+                    // This means the channel is closed
+                    _trace.TraceError("Receiving messages from the service bus threw an OperationCanceledException, most likely due to a closed channel.");
+                    return;
+                }
+                catch (MessagingEntityNotFoundException ex)
+                {
+                    try
+                    {
+                        await receiverContext.Receiver.CloseAsync();
+                    }
+                    catch { }
+
+                    receiverContext.OnError(ex);
+                    var _ = TryCreateSubscription(receiverContext);
```

Comment:
I thought `_` is a quite common convention (go, Swift and EF uses it) but I can change.